### PR TITLE
Begin use of EffectiveConfig Sections in creation of virt backends

### PR DESCRIPTION
**Pull request by cnsnyder**
This PR finishes laying the ground work for using the new ConfigSections to create virt backends.

Please note: As there are no backend specific ConfigSections yet (especially including any that handle encrypted password bits) there are a number of tests which do not pass and will not pass.

I fully expect all tests to pass when the config_refactor branch is merged to master but for this step I think it's necessary for them not to (in order to keep the PRs at least somewhat small).

For testing purposes I expect the libvirt backend to work as expected in all forms of configuration. I've added a temporary commit to modify the libvirt backend to work with a generic config section. Encrypted passwords will still not work with the libvirt backend.

### CODE REVIEW

**Comment by cnsnyder** [comment]
[[8] [CONFIG_EPIC] Create/revise Factory class](https://trello.com/c/hJ3P6WHB/474-8-configepic-create-revise-factory-class)

File: virtwho/config.py [hunk 1]
```diff
@@ -605,13 +573,6 @@ def _readConfig(self, parser):
...
-    def readFile(self, filename):
-        parser = StripQuotesConfigParser()
-        fname = parser.read(filename)
-        if len(fname) == 0:
-            self.logger.error("Unable to read configuration file %s", filename)
-        self._readConfig(parser)
-
     @staticmethod
     def map_destinations_to_sources(configs, dest_classes=(Satellite5DestinationInfo, Satellite6DestinationInfo)):
         """
```
Comment (jirihnidek) [inline virtwho/config.py:578]: Doc string should be updated: `@rtype: (set, set, dict, set)`

File: virtwho/config.py [hunk 2]
```diff
@@ -799,7 +776,7 @@ def iteritems(self):
             yield (key, self._values[key])
 
     def items(self):
-        return [(key, self[key]) for key in self]
+        return sorted([(key, self[key]) for key in self])
```
Comment (jirihnidek) [inline virtwho/config.py]: Method `items` should be iterator to be compatible with Python3 and to bee little bit more memory effective.

File: virtwho/config.py [hunk 3]
```diff
@@ -1096,51 +1104,54 @@ def filter_parameters(desired_parameters, values_to_filter):
...
             if conf.startswith('.'):
                 continue
-            try:
-                filename = parser.read(os.path.join(VW_CONF_DIR, conf))
-                if len(filename) == 0:
-                    logger.error("Unable to read configuration file %s", conf)
-            except MissingSectionHeaderError:
-                logger.error("Configuration file %s contains no section headers", conf)
-        return parser._sections
+            all_sections.update(parse_file(os.path.join(config_dir, conf), parser=parser))
```
Comment (jirihnidek) [inline virtwho/config.py]: The default parses is ` StripQuotesConfigParser` so the `parses=parser` could be omitted here.

File: virtwho/config.py [hunk 4]
```diff
@@ -1096,51 +1104,54 @@ def filter_parameters(desired_parameters, values_to_filter):
...
+    def all_drop_dir_config_sections(config_dir=VW_CONF_DIR):
         """
         Read all configuration sections in the default config directory
-        :return: a dictionary of {section_name: {key: value, ... } ... }
+        :return: a dictionary of {name: {key: value, ... } ... }
         """
         parser = StripQuotesConfigParser()
         all_dir_content = None
         conf_files = None
         non_conf_files = None
```
Comment (jirihnidek) [inline virtwho/config.py]: Initialization of `all_dir_content', 'conf_files' and `non_conf_files` is useless.

File: virtwho/config.py [hunk 5]
```diff
@@ -1075,10 +1069,24 @@ def validate(self):
             # This next check will not be necessary after we know that all sections are
             # ConfigSections
             if getattr(section, 'validate', None) is not None:
-                section.validate(validation_messages=validation_messages)
+                validation_messages.extend(section.validate())
         self.validation_messages = validation_messages
         return validation_messages
 
+    def validation_messages_as_string(self):
```
Comment (jirihnidek) [inline virtwho/config.py]: It is not used at any place and any unit test doesn't exist for this method either.

File: virtwho/config.py [hunk 6]
```diff
@@ -837,17 +814,25 @@ def __len__(self):
...
+        if 'virttype' in self._values:
+            self._values['type'] = self._values['virttype']
         # Finally calls _update_state
         self._update_state()
         return validation_messages
 
     def is_default(self, key):
         return self.defaults[key] == self._values[key]
 
+    def is_section_default(self):
```
Comment (jirihnidek) [inline virtwho/config.py:838]: The name of method could be more self-descriptive or please add doc string.

File: virtwho/config.py [hunk 7]
```diff
@@ -691,18 +651,35 @@ def parse_dests_from_dict(dict_to_parse,
...
-        logger.error("Unable to read configuration file %s", filename)
-    sections = parser._sections
+    parser = parser or StripQuotesConfigParser()
+    sections = {}
+    try:
+        fname = parser.read(filename)
+        if len(fname) == 0:
+            logger.error("Unable to read configuration file %s", filename)
+        else:
+            sections = parser._sections
```
Comment (jirihnidek) [inline virtwho/config.py]: Is it really necessary to use `parses._sections`? I think it is possible to use `parses.sections()` and `parser.remove_section()` at line 708.

File: virtwho/config.py [hunk 8]
```diff
@@ -1204,23 +1226,44 @@ def init_config(env_options, cli_options):
...
     global _effective_config
     _effective_config = effective_config
     effective_config.validate()
 
+    # Log pending errors
+    for err in validation_errors:
+        method = getattr(logger, err[0])
+        if method is not None:
+            method(err[1])
+
```
Comment (jirihnidek) [inline virtwho/config.py]: This looks like familiar to me ;-).

File: virtwho/config.py [hunk 9]
```diff
@@ -1204,23 +1226,44 @@ def init_config(env_options, cli_options):
...
+        validation_errors.append(('warning',
+                                  'Using default "%s" configuration' % VW_ENV_CLI_SECTION_NAME))
+        # In order to keep compatibility with older releases of virt-who,
+        # fallback to using libvirt as default virt backend
+        # only if we did not have a non_default env/cmdline config
+        if not has_non_default_env_cli:
+            effective_config[VW_ENV_CLI_SECTION_NAME] = ConfigSection.from_dict(
+                    DEFAULTS[VW_ENV_CLI_SECTION_NAME],
+                    VW_ENV_CLI_SECTION_NAME,
+                    effective_config)
```
Comment (jirihnidek) [inline virtwho/config.py]: I hope this new code could be in some function.

File: virtwho/config.py [hunk 10]
```diff
@@ -1236,3 +1279,8 @@ def get_effective_config():
     # We don't have one, let us make one
     _effective_config = EffectiveConfig()
     return _effective_config  # Initialize an EffectiveConfig with no options
+
+
+def reset_effective_config():
+    global _effective_config
+    _effective_config = None
```
Comment (jirihnidek) [inline virtwho/config.py]: BTW: I would modify `EffectiveConfig` to be singleton. Using `global` is not my favorite approach. ;-)

**Review by jirihnidek — CHANGES_REQUESTED** [review]
I think that it is on good way. I have some suggestions.

File: virtwho/config.py (hunk 9 above)
Comment (cnsnyder) [reply virtwho/config.py]: I'm not sure I understand? What is the request here?

File: virtwho/config.py (hunk 8 above)
Comment (cnsnyder) [reply virtwho/config.py]: ;-)

**Comment by cnsnyder** [comment]
I have made the EffectiveConfig class a singleton through the use of a metaclass (/me cringes at the thought). Spent quite a bit of time playing with different options for accomplishing this and it seems this works the best for our use case. (Also updated to address the other comments).

Worth mentioning, in our case (thanks to pickle) we can still end up with different instances of the EffectiveConfig.

I look forward to hearing your thoughts.

Cheers!

File: virtwho/config.py (hunk 9 above)
Comment (jirihnidek) [reply virtwho/config.py]: I'm sorry. I meant the block of code between lines 1231-1252 could be refactored into some function (e.g. named `check_valid_conf()`). It is not good practice to have long functions (due to worse readability, etc.). The function `init_config()` is quite long now and it is not worthwhile to make it longer :-).

**Review by jirihnidek — CHANGES_REQUESTED** [review]
I have only one comment containing answer to your question. Overall your other changes are good. 👍

**Review by jirihnidek — APPROVED** [review]
Ack :-)

**Pull request merged**